### PR TITLE
Support `create table` with explicit column definitions

### LINKED ISSUE #4183: Supporting specifying schema when create tables
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I tried `create table t1(a int, b int, c string) as values(0, 100, 'a')` but failed.
The response is ``NotImplemented("Only `CREATE TABLE table_name AS SELECT ...` statement is supported")``.

**Describe the solution you'd like**
Supporting specifying schema when create a table.

### PR DESCRIPTION
**Pull request by doki23**
# Which issue does this PR close?
Closes #4183 

# What changes are included in this PR?
Most changes are in planner of creating memory table.

# Are these changes tested?
Yes.

# Are there any user-facing changes?
Users can create a table with a specified schema with a values query.

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -180,12 +180,36 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                                cast(col(input_field.name()), field.data_type().clone()).alias(field.name())
+                            }).collect::<Vec<_>>();
+                            LogicalPlanBuilder::from(plan.clone())
+                                .project(project_exprs)?
+                                .build()?
+                        },
+                        _ => return Err(DataFusionError::Plan(
+                            "You can only specify schema when create table with a `values` statement"
+                                .to_string()
+                        ))
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:208]: I wonder why not support queries as well? Why only `VALUES` -- the code you have written should work for a `CREATE TABLE AS SELECT` as well

**Review by alamb — COMMENTED** [review]
Thank you @doki23  -- this looks like a nice improvement! I have a few suggestions -- the only thing I think is needed prior to merge is a test for mismatched column count. 

We can file a follow up to support specifying column types from a create table as select as a follow on PR

File: datafusion/sql/src/planner.rs (hunk 1 above)
Comment (doki23) [reply datafusion/sql/src/planner.rs:208]: It's because `SELECT` has its own schema, but it makes sense if someone want to cast these types.

**Comment by doki23** [comment]
> Thank you @doki23 -- this looks like a nice improvement! I have a few suggestions -- the only thing I think is needed prior to merge is a test for mismatched column count.
> 
> We can file a follow up to support specifying column types from a create table as select as a follow on PR

I'm going to do it in this pr, here's no need to file a follow up issue.

File: datafusion/core/tests/sql/create_drop.rs [hunk 2]
```diff
@@ -49,6 +49,76 @@ async fn create_table_as() -> Result<()> {
...
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn create_table_with_schema_as_select_mismatch() -> Result<()> {
+    let ctx = SessionContext::new();
+    register_aggregate_simple_csv(&ctx).await?;
+
+    let sql = "CREATE TABLE my_table(c1 float, c2 double, c3 boolean, c4 varchar) \
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:82]: 👍

File: datafusion/core/tests/sql/create_drop.rs [hunk 3]
```diff
@@ -49,6 +49,76 @@ async fn create_table_as() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn create_table_with_schema_as_select() -> Result<()> {
+    let ctx = SessionContext::new();
+    register_aggregate_simple_csv(&ctx).await?;
+
+    let sql = "CREATE TABLE my_table(c1 float, c2 double, c3 boolean, c4 varchar) \
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:57]: its-so-beautiful-crying-gif

**Review by alamb — APPROVED** [review]
Thank you @doki23  -- this looks great!

**Comment by alamb** [comment]
I took the liberty of merging this PR up to master and fixing the clippy issue so we can get it merged. Thanks again @doki23

**Comment by alamb** [comment]
BTW I also made a small change to one of the tests to satisfy clippy https://github.com/apache/arrow-datafusion/pull/4194/commits/a2172db162e26c2f729663d541f337db3b006ae9

**Comment by alamb** [comment]
Thanks again @doki23

**Pull request merged**